File: internal/cipher/cipher_test.go
package cipher

import (
	"bytes"
	"compress/gzip"
	"testing"
)

func TestSealAndOpen(t *testing.T) {
	input := []byte("gzip-like bytes")
	sealed := SealWithChecksum(input)
	opened, ok := OpenWithChecksum(sealed)

	if len(sealed) != len(input)+ChecksumBytes {
		t.Fatalf("sealed length = %d, want %d", len(sealed), len(input)+ChecksumBytes)
	}
	if !ok || !bytes.Equal(opened, input) {
		t.Fatalf("opened = %q, want %q", opened, input)
	}
}

func TestOpenChecksumMismatch(t *testing.T) {
	sealed := SealWithChecksum([]byte("payload"))
	sealed[0] ^= 0xff

	if _, ok := OpenWithChecksum(sealed); ok {
		t.Fatal("expected checksum mismatch")
	}
}

func TestCRC32Stable(t *testing.T) {
	if got := CRC32([]byte("test")); got != 0xd87f7e0c {
		t.Fatalf("crc32 = %#x, want %#x", got, 0xd87f7e0c)
	}
}

func TestXorBytes(t *testing.T) {
	var buf bytes.Buffer
	gz := gzip.NewWriter(&buf)
	gz.Write([]byte("hello world"))
	gz.Close()
	input := buf.Bytes()

	output := XorBytes(input, DefaultKey)
	if len(output) != len(input) {
		t.Fatalf("xor length = %d, want %d", len(output), len(input))
	}

	encrypted := XorBytes([]byte("test payload"), "secret")
	decrypted := XorBytes(encrypted, "secret")
	if !bytes.Equal(decrypted, []byte("test payload")) {
		t.Fatalf("decrypted = %q", decrypted)
	}

	obfuscated := XorBytes(input, DefaultKey)
	if bytes.Equal(obfuscated, input) {
		t.Fatal("expected obfuscated output to differ from input")
	}
}

File: test/cipher.test.js
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { gzipSync } from "node:zlib";
import { CHECKSUM_BYTES, crc32, DEFAULT_KEY, openWithChecksum, sealWithChecksum, xorBytes } from "../lib/cipher.js";

describe("checksum", () => {
  it("seals and opens a payload", () => {
    const input = Buffer.from("gzip-like bytes");
    const sealed = sealWithChecksum(input);
    const opened = openWithChecksum(sealed);

    assert.equal(sealed.length, input.length + CHECKSUM_BYTES);
    assert.deepEqual(opened, input);
  });

  it("returns null when checksum does not match", () => {
    const sealed = sealWithChecksum(Buffer.from("payload"));
    sealed[0] ^= 0xff;

    assert.equal(openWithChecksum(sealed), null);
  });

  it("computes a stable crc32", () => {
    assert.equal(crc32(Buffer.from("test")), 0xd87f7e0c);
  });
});

describe("xorBytes", () => {
  it("preserves byte length", () => {
    const input = gzipSync(Buffer.from("hello world"));
    const output = xorBytes(input, DEFAULT_KEY);

    assert.equal(output.length, input.length);
  });

  it("is reversible with the same key", () => {
    const input = Buffer.from("test payload");
    const encrypted = xorBytes(input, "secret");
    const decrypted = xorBytes(encrypted, "secret");

    assert.deepEqual(decrypted, input);
  });

  it("changes output without changing size", () => {
    const input = gzipSync(Buffer.from("<p>hello</p>"));
    const obfuscated = xorBytes(input, DEFAULT_KEY);

    assert.notDeepEqual(obfuscated, input);
    assert.equal(obfuscated.length, input.length);
  });
});
